src/librustc_typeck/astconv.rs

```diff
@@ -970,60 +970,79 @@ impl<'o, 'gcx: 'tcx, 'tcx> dyn AstConv<'gcx, 'tcx> + 'o {
     {
         let tcx = self.tcx();
 
-        if trait_bounds.is_empty() {
-            span_err!(tcx.sess, span, E0224,
-                      "at least one non-builtin trait is required for an object type");
-            return tcx.types.err;
-        }
-
         let mut projection_bounds = Vec::new();
+        let mut potential_assoc_types = Vec::new();
         let dummy_self = self.tcx().types.trait_object_dummy_self;
-        let (principal, potential_assoc_types) = self.instantiate_poly_trait_ref(
-            &trait_bounds[0],
-            dummy_self,
-            &mut projection_bounds,
-        );
-        debug!("principal: {:?}", principal);
-
-        for trait_bound in trait_bounds[1..].iter() {
-            // sanity check for non-principal trait bounds
-            self.instantiate_poly_trait_ref(trait_bound,
-                                            dummy_self,
-                                            &mut vec![]);
+        let bound_trait_refs: Vec<_> = trait_bounds
+            .iter()
+            .rev()
+            .map(|trait_bound| {
+                let (trait_ref, cur_potential_assoc_types) = self.instantiate_poly_trait_ref(
+                    trait_bound,
+                    dummy_self,
+                    &mut projection_bounds
+                );
+                potential_assoc_types.extend(cur_potential_assoc_types.into_iter().flatten());
+                (trait_ref, trait_bound.span)
+            })
+            .collect();
+
+        // Expand trait aliases recursively and check that only one regular (non-auto) trait
+        // is used and no 'maybe' bounds are used.
+        let expanded_traits = traits::expand_trait_aliases(tcx, bound_trait_refs.clone());
```

Comment:
`expand_trait_aliases` is only ever called for `dyn TraitAlias`, right? If so, the code in `expand_trait_aliases` could fetch the `Hir` of any trait alias and check for `?Trait` bounds and error on them directly.

Reply:
Hmm, good point. (And yes, `expand_trait_aliases` is a fn I introduced in this PR specifically for that purpose.) Let me try it out.

Reply:
Okay, I did actually consider this before, but the issue is: you can't get the HIR for a cross-crate trait alias.

Reply:
ugh. Right. I guess we need to figure out a way to return this information from `super_predicates_of`. I haven't looked into it, but can you check out how that would be doable and give us some options to choose from along with pros/cons?

Reply:
Indeed. I sort of hoped to avoid that, since it will involve a lot more changes to code, but oh well.

Reply:
Indeed. I sort of hoped to avoid that, since it will involve a lot more changes to code, but oh well.

Reply:
Okay, here are some options @oli-obk:
* Modify the `Predicate<'tcx>` enum so that the `Trait(PolyTraitPredicate<'tcx>)` variant becomes something like `Trait(PolyTraitPredicate<'tcx>, BoundKind)`.
* Add a flag to the `TraitPredicate<'tcx>` struct (which right now just contains `pub trait_ref: TraitRef<'tcx>`). This is similar to the previous solution, but a bit finer-grained, I suppose.
* Extend the `GenericPredicates` struct to include a flag for whether a `?Sized` bound was encountered (alternatively a set of maybe bounds encountered, though obviously `?Sized` is the only one right now).

I'm not sure what exactly gets (or can easily get) serialised into crate metadata though.

Reply:
Thanks.

> I'm not sure what exactly gets (or can easily get) serialised into crate metadata though.

Since they are returned by a query, all of these are automatically serialized.

All of these three solutions seem somewhat hacky. I don't see a good other way though.

One thing I'm wondering is that if one crate exposes `fn foo<T: ?Sized>(t: &T);`, how does another crate know it is allowed to pass a `dyn` type for `T`? This information must be available somewhere, even if the function is defined as `fn foo<T: TraitAlias>(t: &T)`.

Reply:
Right, makes sense. I seem to recall that from previous work now.

They're all a little hacky, but probably not *too* bad I reckon. Regarding your point about `T: ?Sized`, that's interesting... have you tested it already, and does it work as expected for `T: TraitAlias`?

Reply:
One option is to simply ban `?Sized` pre-expansion; this means that stable behavior doesn't change and we can consider changes later depending on how hacky they become. I think that's serviceable so that we get this fix off the ground. I believe the pre-expansion banning was already implemented.

Reply:
Well, at least that's easy... definitely a stop-gap solution, but let's see.

Reply:
@oli-obk Any more thoughts?

Reply:
Yea, banning them pre-expansion seems reasonable right now, since it's the most restrictive variant, keeping us all options open.

Reply:
@oli-obk Well not quite. It's a shallow ban, which means you can still use them if they're within trait aliases (it will silently ignore).

Reply:
Huh? I thought @Centril meant to just ban `?Trait` bounds on the HIR of trait aliases? What exactly are we banning? Do you have an example code?

Reply:
@oli-obk I meant specifically in `dyn T0 + ... + Tn`, not inside trait aliases (which are unstable in any case). This is what the code does right now.

Reply:
@oli-obk I think we may want to just implement one of my three suggestions above for now, if you're not a fan of @centril's suggestion (I'm open)... let me know what you prefer.

Reply:
I think I just figured out how it works. If you have no `?Sized` bound, the `Predicate` list will contain `Sized` as a `Predicate::Trait`. If you have a `?Sized` bound, there will be no such entry. So you can do this cross crate already by checking whether there is a `Sized` predicate, and if not, error.

This means that `?Sized` bounds are allowed in trait aliases, but that makes the trait alias unusable for `dyn` trait objects. It is a little counterintutitive, that `dyn` trait objects only allow `Sized` trait aliases, but that's what we are trying to enforce here, right?

Reply:
@oli-obk Yep, that's exactly what we're trying to enforce. What you say makes sense actually... I could have sworn I had tested for that behaviour previously, but evidently not! The question is: do we want to simply test for overall presence of `Sized` among predicates, or test individually during trait alias expansion? That is, should the following be accepted or error?

```rust
trait A = Foo + Send + ?Sized;
trait B = Sync; // implicitly included `Sized` predicate
type T = dyn A + B; // does this complain about a `?Sized` bound or not?
```

Reply:
Intuitively I'd have said that code example would not error, but we can start out with it erroring and relax that later when we are sure it's allright.

Reply:
Okay. Are you sure? Just want to avoid flip-flopping before I code this. Especially since permitting the above would definitely be easier to implement. :-)

Reply:
I'll leave this to @Centril / T-lang, I have no opinion or good knowledge of the implications.

Reply:
Let's just ban it shallowly for now and possibly reconsider later as a smaller step if necessary. Trait aliases aren't getting stabilized soon in any case. Just leave a note on the tracking issue or something.